components/motor/ulnstepper/28byj-48.go

```diff
@@ -0,0 +1,478 @@
+	}
+
+	if mc.Pins.In1 != "" {
+		in1, err := b.GPIOPinByName(mc.Pins.In1)
+		if err != nil {
+			return nil, err
+		}
+		m.in1 = in1
+	}
+
+	if mc.Pins.In2 != "" {
+		in2, err := b.GPIOPinByName(mc.Pins.In2)
+		if err != nil {
+			return nil, err
+		}
+		m.in2 = in2
+	}
+
+	if mc.Pins.In3 != "" {
+		in3, err := b.GPIOPinByName(mc.Pins.In3)
+		if err != nil {
+			return nil, err
+		}
+		m.in3 = in3
+	}
+
+	if mc.Pins.In4 != "" {
+		in4, err := b.GPIOPinByName(mc.Pins.In4)
+		if err != nil {
+			return nil, err
+		}
+		m.in4 = in4
+	}
+
+	if err := m.Validate(); err != nil {
+		return nil, err
+	}
+
+	m.startThread(ctx)
+	return m, nil
+}
+
+// struct is named after the controler uln2003.
+type uln2003 struct {
+	theBoard           board.Board
+	rotationPerMinute  float64
+	ticksPerRotation   int
+	in1, in2, in3, in4 board.GPIOPin
+	logger             golog.Logger
+	motorName          string
+
+	// state
+	lock  sync.Mutex
+	opMgr operation.SingleOperationManager
+
+	stepPosition         int64
+	threadStarted        bool
+	targetStepPosition   int64
+	targetStepsPerSecond int64
+	generic.Unimplemented
+}
+
+// validate if this config is valid.
+func (m *uln2003) Validate() error {
+	if m.theBoard == nil {
+		return errors.Errorf("need a board for motor (%s)", m.motorName)
+	}
+
+	if m.rotationPerMinute == 0 {
+		return errors.Errorf("need to set 'rotation_per_minute' for motor (%s)", m.motorName)
+	}
+
+	if m.in1 == nil {
+		return errors.Errorf("need a 'In1' pin for motor (%s)", m.motorName)
+	}
+
+	if m.in2 == nil {
+		return errors.Errorf("need a 'In2' pin for motor (%s)", m.motorName)
+	}
+
+	if m.in3 == nil {
+		return errors.Errorf("need a 'In3' pin for motor (%s)", m.motorName)
+	}
+
+	if m.in4 == nil {
+		return errors.Errorf("need a 'In4' pin for motor (%s)", m.motorName)
+	}
+
+	return nil
+}
+
+// SetPower sets the percentage of power the motor should employ between 0-1.
+func (m *uln2003) SetPower(ctx context.Context, powerPct float64, extra map[string]interface{}) error {
+	if math.Abs(powerPct) <= .0001 {
+		m.stop()
+		return nil
+	}
+	return errors.Errorf("doesn't support raw power mode in motor (%s)", m.motorName)
+}
+
+func (m *uln2003) startThread(ctx context.Context) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	if m.threadStarted {
+		return
+	}
+
+	m.threadStarted = true
+	go m.doRun(ctx)
+}
+
+// setStepperDelay calculates the wait time between each step taken.
+// The minimum value is set to 0.002s, anything less then this can potentially damage the gears.
+func (m *uln2003) setStepperDelay(rpm float64) float64 {
+	stepperDelay := 1 / ((rpm * 8) / 60)
+	return math.Max(stepperDelay, minStepDelay)
+}
+
+func (m *uln2003) doRun(ctx context.Context) {
+	for {
+		sleep, err := m.doCycle(ctx)
+		if err != nil {
+			m.logger.Info("error in uln2003 %w", err)
+		}
+
+		if !utils.SelectContextOrWait(ctx, sleep) {
+			return
+		}
+	}
+}
+
+func (m *uln2003) doCycle(ctx context.Context) (time.Duration, error) {
```

Comment:
This function is doing two different things:
1. performing the actual step sequence
2. returning an amount of time to sleep before the next step sequence

The code is easier to follow if we handle the sleep conditions within `doRun` itself and remove `doCycle` completely because the function running the for-loop is now in charge of the sleep time for each iteration (and because there are less links in our chain of functions)

Reply:
merged doCycle into doRun